### ISSUE TITLE
Try out include-what-you-use

### ISSUE THREAD

**Issue by cameel** [ethereum/solidity#11813]
This is a part of #10269 but getting includes consistent also came up recently in #11812 as it might prevent bugs like #11811 and #11103. I'm creating a more focused issue to try out specifically [include-what-you-use](https://include-what-you-use.org) even if it does not solve the original problem as a whole.

---

The task here is to install [include-what-you-use](https://include-what-you-use.org) locally and run it on our codebase. Then submit changes in a PR so that we can discuss them and decide if it's good enough to add it as a PR check that runs automatically (it could run as a part of the [`chk_coding_style` job](https://github.com/ethereum/solidity/blob/v0.8.7/.circleci/config.yml#L320-L336)). Please also post all the commands you used for installing/building/running it.

The main criterion is whether it's reliable enough and does not cause too many false positives that would have to be ignored in some way or, worse, would make the check fail on perfectly valid PRs. Please also make sure that it works not only with our own headers but also with ones from STL and other libraries.

**Comment by akbhuker** [comment]
hey, can I have this issue?
I'm quite new to this so I might need some guidance

**Comment by cameel** [comment]
Sure, go ahead!

If you need help, you can always ask in this issue or on the `#solidity-dev` [Matrix](https://app.element.io/#/room/#ethereum_solidity-dev:gitter.im)/[Gitter](https://gitter.im/ethereum/solidity-dev) channel.

**Comment by Sild** [comment]
Hi, I went around this issue

TLDR: it does't work from the box: solidity does not compile after iwyu updates (some of includes are broken I guess)

Reproduction snippet:

```
cd ~
apt install libclang-13-dev
git clone https://github.com/include-what-you-use/include-what-you-use.git
cd include-what-you-use
git fetch --all && git checkout clang_13
mkdir build && cd build
cmake -G "Unix Makefiles" -DCMAKE_C_COMPILER=/usr/bin/clang-13 -DCMAKE_CXX_COMPILER=/usr/bin/clang++-13 -DCMAKE_PREFIX_PATH=/usr/lib/llvm-13/ ../
make
mv bin/include-what-you-use /usr/bin/
cd /solidity/build
cmake -DCMAKE_C_COMPILER=/usr/bin/clang-13 -DCMAKE_CXX_COMPILER=/usr/bin/clang++-13 -DUSE_Z3=OFF -DCMAKE_CXX_INCLUDE_WHAT_YOU_USE=include-what-you-use ../
# build is really slow with iwyu
cmake --build . > build_output.txt 2>&1
cat build_output.txt | grep -Ev "\[\ +[0-9]+" > iwyu_output.txt
python3 ~/include-what-you-use/fix_includes.py < iwyu_output.txt
mv *output.txt ~/ && rm -rf *
cmake -DCMAKE_C_COMPILER=/usr/bin/clang-13 -DCMAKE_CXX_COMPILER=/usr/bin/clang++-13 -DUSE_Z3=OFF ../ 
cmake --build .
```

What is the next expectation? Should I try to fix build and provide the correct config for `iwyu` here?

**Comment by axic** [comment]
@Sild can you share a commit with the changes? I'd be curious to see what changes iwyu did.

**Comment by Sild** [comment]
@axic sure https://github.com/ethereum/solidity/pull/13589

**Comment by Sild** [comment]
after cleanup iwyu_output from boost and bits additional includes, it's become better

```
cat iwyu_output.txt | grep -v "boost\|bits" > iwyu_output_no_boost_no_bits.txt
```

but linker still fail somewhere
```
[ 56%] Linking CXX executable solidity-upgrade
/solidity/libsolidity/ast/AST.h:0: error: undefined reference to 'typeinfo for solidity::frontend::FunctionDefinition'
/solidity/libsolidity/ast/AST.h:0: error: undefined reference to 'typeinfo for solidity::frontend::FunctionDefinition'
/usr/bin/../lib/gcc/aarch64-linux-gnu/11/../../../../include/c++/11/bits/vector.tcc:504: error: undefined reference to 'typeinfo for solidity::frontend::FunctionDefinition'
...
...
```

**Comment by cameel** [comment]
Just for some extra context since I see it wasn't posted here: we did try include-what-you-use and the results weren't very promising in terms of having it as a fire-and-forget PR check https://github.com/ethereum/solidity/issues/10269#issuecomment-1062802540.

**Comment by Sild** [comment]
I've got the same result at first.

Then possible it's better to close this issue?

**Comment by cameel** [comment]
Right. I guess the work is done here. Thanks for confirming though!
